### ISSUE TITLE
Add new rules for better testing experience

### ISSUE THREAD

**Issue by elaichenkov** [playwright-community/eslint-plugin-playwright#43]
I'd love to have such rules in the plugin. Moreover, I believe that I have the capacity to implement them. 
@mxschmitt @mskelton What do you think?

| Default  | Level    | Rule name                    | Description          |      
| :----:                         | :----:       |    :----                          |           :----            |
| ✅                              | ❌ error | [no-focused-test](https://github.com/playwright-community/eslint-plugin-playwright#no-focused-test)         | Disallow focused tests                      |
| ✅                              | ❌ error | no-identical-test-title | Disallow identical test titles              |
| ✅                              | ❌ error | require-assertion       | Require assertion in test                   |
| ✅                              | ⚠️ warn  | [no-force-option](https://github.com/playwright-community/eslint-plugin-playwright#no-force-option)         | Disallow using `{ force: true }` option     |
| ✅                              | ⚠️ warn  | [no-wait-for-timeout](https://github.com/playwright-community/eslint-plugin-playwright#no-wait-for-timeout)     | Disallow using `page.waitForTimeout` method |
|                                   | ❌ error | no-serial-test          | Disallow serial tests                       |
|                                   | ⚠️ warn  | [no-skipped-test](https://github.com/playwright-community/eslint-plugin-playwright#no-skipped-test)         | Disallow skipped tests                      |
|                                   | ⚠️ warn | prefer-to-have-describe | Prefer including test inside describe block |
|                                   | ❌ error | no-xpath-selector       | Disallow xPath selectors                    |
|                                   | ❌ error | no-text-selector        | Disallow text selectors                     |

**Comment by mskelton** [comment]
Sounds fine with me, I like those proposals.  Definitely don't think they should all be default rules, but I like having options for users.

**Comment by mxschmitt** [comment]
I think we can add these:
- no-focused-test
- require-assertion
- no-force-option
- no-wait-for-timeout
- no-skipped-test


Not sure about `no-serial-test` probably good to at least warn about it, since its discouraged.  

We don't need `no-identical-test-title` because the test-runner enforces it already.

Not sure about `prefer-to-have-describe` since we at Playwright e.g. try to do the opposite, so its more flat and prefer moving similar tests out to separate files.

`no-xpath-selector` and `no-text-selector` as long they are not default, I'm fine with them. I just think having a text selector is fine? But test-ids e.g. are for sure better.

**Comment by elaichenkov** [comment]
> Sounds fine with me, I like those proposals. Definitely don't think they should all be default rules, but I like having options for users.

Sure. I added the `default` column. Let me know your thoughts.

> Not sure about `no-serial-test` probably good to at least warn about it, since its discouraged.

We can make it `warn`.


> We don't need `no-identical-test-title` because the test-runner enforces it already.

Yeah, I know that. But ESlint will show this error statically before running the tests. So, I'm pretty sure that would be extremely valuable to users.

> Not sure about `prefer-to-have-describe` since we at Playwright e.g. try to do the opposite, so its more flat and prefer moving similar tests out to separate files.

I see. In my view, a lot of users group tests together using a `describe` block. It's widely used for better organizing.

> `no-xpath-selector` and `no-text-selector` as long they are not default, I'm fine with them. I just think having a text selector is fine? But test-ids e.g. are for sure better.

Yep, they would never be default. Because people have to choose to go on their own their selector strategies. We can remove the `no-text-selector` rule from the list. However, I think that using just text as a selector is really fragile. Because text content may change.

**Comment by elaichenkov** [comment]
Btw, I'm certain that we need to add something like this table in the README file. How do you feel about this

**Comment by unlikelyzero** [comment]
@elaichenkov i think the text select would be fine since it's used heavily in screen readers, no?

**Comment by mskelton** [comment]
@unlikelyzero I agree that testing by text is fine, and in my opinion it's the preferred approach whenever possible.  Having it as an option though if a user wants it is fine, just not enabled by default.

**Comment by elaichenkov** [comment]
> @elaichenkov i think the text select would be fine since it's used heavily in screen readers, no?

Yeah, that's fine. But it's really better to use `data-test-*` attributes whenever possible. Perhaps, we should consider adding the `require-selectors` rule, which will allow using an array of selectors (e.g. `['test-id', 'id', 'text']`). This rule would be flexible and extensible.

**Comment by elaichenkov** [comment]
More than that, probably we should add the `max-selector-length` rule, which enforces maximum selector lengths. This rule can help to avoid long `css` or `xpath` chains. 

For instance:
❌ Example of **incorrect** usage:
```js
/* "playwright/max-selector-length": ["error", { "max": 25 }] */
await page.locator('#tsf > div:nth-child(2) > div.A8SBwf > div.RNNXgb > div > div.a4bIc > input').click();
```
✅ Example of **correct** usage:
```js
/* "playwright/max-selector-length": ["error", { "max": 25 }] */
await page.locator('#tsf > div:nth-child(2)').click();
```

**Comment by mskelton** [comment]
> But it's really better to use data-test-* attributes whenever possible.

I strongly disagree, and want to make sure this plugin does not project that opinion.  There are different approaches to this and I strongly believe the best approach is testing as close to user behavior as possible which means testing user facing content whenever possible (text, aria-labels, etc.)

Testing Library explains these principles really well.  https://testing-library.com/docs/queries/about#priority

**Comment by elaichenkov** [comment]
> I strongly disagree, and want to make sure this plugin does not project that opinion. There are different approaches to this and I strongly believe the best approach is testing as close to user behavior as possible which means testing user facing content whenever possible (text, aria-labels, etc.)
> 
> Testing Library explains these principles really well. https://testing-library.com/docs/queries/about#priority

I see. That's interesting and thanks for sharing that. However, I just want to notice that comments/opinions are my own and not the views of this project. Moreover, as I said I believe that everyone should choose their own test and selector strategies as well. The plugin should provide a set of rules which users can enable/disable for their needs.

**Comment by elaichenkov** [comment]
Any thoughts on all other proposals?

**Comment by mskelton** [comment]
> From my point of view, the plugin should provide a bunch of rules which users can enable/disable for their needs

Yep, I completely agree 🙂  ESLint has rules that I would never use, but other users may find them invaluable.  It's all about giving options to users and enabling rules that are universally accepted as recommend for the defaults.

**Comment by unlikelyzero** [comment]
> Any thoughts on all other proposals?

This covers all the applicable rules from the cypress plugin at least

**Comment by elaichenkov** [comment]
What about `require-selectors` and `max-selector-length`? Are these really needed?

**Comment by unlikelyzero** [comment]
> What about `require-selectors` and `max-selector-length`? Are these really needed?

I've been on a couple of teams which could have used these.

**Comment by mxschmitt** [comment]
Just had the idea that we could add a rule e.g. dont-use-page-methods since page.* methods have no strict mode enabled by default, only locators have.

**Comment by elaichenkov** [comment]
> Just had the idea that we could add a rule e.g. dont-use-page-methods since page.* methods have no strict mode enabled by default, only locators have.

Sounds reasonable. Just to clarify that you mean the rule should disallow using methods such as:
```
page.$
page.$$
page.$$eval
page.$eval
page.check
page.click
page.close
page.dblclick
page.dragAndDrop
page.evaluate
page.evaluateHandle
page.fill
page.focus
page.frame
page.frames
page.getAttribute
page.hover
page.innerHTML
page.innerText
page.inputValue
page.isChecked
page.isClosed
page.isDisabled
page.isEditable
page.isEnabled
page.isHidden
page.isVisible
page.pause
page.press
page.screenshot
page.selectOption
page.setChecked
page.setContent
page.setInputFiles
page.tap
page.textContent
page.type
page.uncheck
```
So, probably all methods which locator class has, right?

**Comment by mskelton** [comment]
Why not page.screenshot? Users can do it with locator.

Taking full page screenshots is perfectly valid and doesn't have the issues that Max mentioned that locators aim to solve (selector strict mode).

**Comment by kkumarlinz** [comment]
> I'd love to have such rules in the plugin. Moreover, I believe that I have the capacity to implement them. @mxschmitt @mskelton What do you think?
> 
> Default	Level	Rule name	Description
> ✅	❌ error	[no-focused-test](https://github.com/playwright-community/eslint-plugin-playwright#no-focused-test)	Disallow focused tests
> ✅	❌ error	no-identical-test-title	Disallow identical test titles
> ✅	❌ error	require-assertion	Require assertion in test
> ✅	⚠️ warn	[no-force-option](https://github.com/playwright-community/eslint-plugin-playwright#no-force-option)	Disallow using `{ force: true }` option
> ✅	⚠️ warn	[no-wait-for-timeout](https://github.com/playwright-community/eslint-plugin-playwright#no-wait-for-timeout)	Disallow using `page.waitForTimeout` method
> ❌ error	no-serial-test	Disallow serial tests
> ⚠️ warn	[no-skipped-test](https://github.com/playwright-community/eslint-plugin-playwright#no-skipped-test)	Disallow skipped tests
> ⚠️ warn	prefer-to-have-describe	Prefer including test inside describe block
> ❌ error	no-xpath-selector	Disallow xPath selectors
> ❌ error	no-text-selector	Disallow text selectors

Waiting for these rules to be implemented sooner

**Comment by mskelton** [comment]
@kkumarlinz We just release a new version this morning with many of these new rules.  Docs coming soon.

**Comment by mskelton** [comment]
Closing in favor of the larger issue related to parity with the Jest plugin.
